src/Decompiler/Analysis/IndirectCallRewriter.cs

```diff
@@ -95,14 +102,57 @@ private void RewriteCall(Statement stm, CallInstruction call)
             var ft = pt.Pointee as FunctionType;
             if (ft == null)
                 return;
-            var returnId = ft.ReturnValue.DataType is VoidType ?
-                null : ft.ReturnValue;
-            var sigCallee = new FunctionType(returnId, ft.Parameters);
+            AdjustStack(stm, call, ft.StackDelta);
             var ab = new FrameApplicationBuilder(
                  program.Architecture, proc.Frame, call.CallSite,
-                 call.Callee, true);
-            stm.Instruction = ab.CreateInstruction(sigCallee, null);
+                 call.Callee, false);
+            stm.Instruction = ab.CreateInstruction(ft, null);
             ssaIdTransformer.Transform(stm, call);
+            changed = true;
+        }
+
+        private void AdjustStack(
+            Statement stm,
+            CallInstruction call,
+            int stackDelta)
+        {
+            var defSpBinding = call.Definitions.Where(
+                u => u.Storage == program.Architecture.StackRegister)
+                .FirstOrDefault();
+            if (defSpBinding == null)
+                return;
+            var defSpId = defSpBinding.Expression as Identifier;
+            if (defSpId == null)
+                return;
+            var usedSpExp = call.Uses.Where(
+                u => u.Storage == program.Architecture.StackRegister)
+                .Select(u => u.Expression)
+                .FirstOrDefault();
+            if (usedSpExp == null)
+                return;
+            var retSize = call.CallSite.SizeOfReturnAddressOnStack;
+            var offset = stackDelta - retSize;
+            var src = emiter.IAdd(usedSpExp, Constant.Word32(offset));
+            var ass = new Assignment(defSpId, src);
```

Comment:
Add comment:
// Generate a statement that adjusts the stack pointer according to the calling convention.